### DISCUSSION TITLE
Why does filesystem::read_symlink ignore junctions?

### DISCUSSION THREAD

**Discussion by operatios** [microsoft/STL#2809] [Q&A] [answered]
I do understand that symbolic links and junctions are different, still, they both point to a target. For example, as a user, I expected to be able to successfully read both `"C:\Users\All Users"` (a symbolic link)  and `"C:\Users\Default User"` (a junction).

After looking at the source code: https://github.com/microsoft/STL/blob/7f04137880e1c3df5125c1baf808f16f399dee2e/stl/src/filesystem.cpp#L515 I see that only `IO_REPARSE_TAG_SYMLINK` is checked, and in case of junctions (tag value `IO_REPARSE_TAG_MOUNT_POINT`), an error `ERROR_REPARSE_TAG_INVALID` is returned. This would be fine if there existed a function like `read_junction`, but as far as I am aware, there is none.

This feels weird to me since other in programming languages like Python, Rust and Go you are able to successfully read both junctions and symbolic links on Windows. Here is a reference to their implementations, if needed: Python's [os.readlink](https://github.com/python/cpython/blob/0efe3a1636c143fe0694a8e4d25d6eae19e0d618/Modules/posixmodule.c#L8692), Rust's [fs::read_link](https://github.com/rust-lang/rust/blob/17c6bde14eabdd61c35505ed6834ee9196c2fdc9/library/std/src/sys/windows/fs.rs#L506) and Go's [os.Readlink](https://github.com/golang/go/blob/19309779ac5e2f5a2fd3cbb34421dafb2855ac21/src/os/file_windows.go#L483).

After doing some more investigation, I found that a similar question was raised as an issue in boost: boostorg/filesystem#125 and later was [fixed](https://github.com/boostorg/filesystem/pull/100). I am just not sure whether this is a bug or not, since libc++'s `read_symlink` also does not support junctions.

Is this something that is not allowed by the C++ Standard, or is there another reason?

**Comment by fsb4000** [comment]
I see the same issue: https://github.com/microsoft/STL/issues/2077
But I don't know why it is closed as invalid...

**Comment by strega-nil-ms** [comment, marked as answer]
@barcharcraz points out that the standard wording for `read_symlink` is ([\[fs.op.read.symlink\]/1][]):

> Returns: If `p` resolves to a symbolic link, a `path` object containing the contents of that symbolic link.

With the definition of "symbolic link" being ([\[fs.general\]/def:symbolic_link][]):

> A symbolic link is a type of file with the property that when the file is encountered during pathname resolution ([\[fs.class.path\]][]), a string stored by the file is used to modify the pathname resolution.

This says to me that junctions _should_ be considered symbolic links, and therefore `read_symlink` and `copy_symlink` should do the obvious thing with junctions as well.

There is a question here, which is whether `copy_symlink` should create a junction or a directory symlink; in my opinion, the correct behavior is to create a junction.

So, to answer your question: I do believe that this is a bug.

[\[fs.op.read.symlink\]/1]: https://timsong-cpp.github.io/cppwp/n4868/fs.op.read.symlink#1
[\[fs.general\]/def:symbolic_link]: https://timsong-cpp.github.io/cppwp/n4868/fs.general#def:symbolic_link
[\[fs.class.path\]]: https://timsong-cpp.github.io/cppwp/n4868/fs.class.path
